concur-core/src/Concur/Core/Types.hs

```diff
@@ -111,20 +118,57 @@ instance Monoid v => Alternative (Widget v) where
 
 instance Monoid v => MultiAlternative (Widget v) where
   never = display mempty
-  orr = Widget . comb . map suspend
+  orr ws = do
+    mvar <- io newEmptyMVar
+    comb mvar $ fmap (Left . step) ws
     where
-      peelAllFree []           = Right []
-      peelAllFree (Pure a : _) = Left a
-      peelAllFree (Free s: fs) = fmap (s:) $ peelAllFree fs
-      comb wfs = case peelAllFree wfs of
-        Left a -> pure a
-        Right fsio -> Free $ Suspend $ do
-          fs <- mapM unSuspend fsio
-          return $ SuspendF (mconcat $ map view fs) (merge $ map cont fs)
-        where
-          merge ws = do
-            (i, me) <- foldl (\prev (i,w) -> prev <|> fmap (i,) w) retry $ zip [0..] ws
-            return $ fmap (\e -> comb $ take i wfs ++ [e] ++ drop (i+1) wfs) me
+      go :: v -> Free (SuspendF v) a -> IO (Either a (v, IO (Free (SuspendF v) a)))
+      go _ (Free (StepView v next))  = go v next
+      go v (Free (StepIO a next))    = a >>= go v . next
+      go v (Free (StepBlock a next)) = pure $ Right (v, a >>= pure . next)
+      go _ (Pure a)                  = pure $ Left a
+  
+      comb :: MVar (Int, Free (SuspendF v) a) -> [Either (Free (SuspendF v) a) (v, ThreadId)] -> Widget v a
+      comb mvar vs = do
+        rs <- io $ forM vs $ \v -> case v of
+          Left w -> do
+            r <- go mempty w
+            pure $ case r of
+              Left a -> Left a
+              Right b -> Right $ Left b
+          Right (w, tid) -> pure $ Right $ Right (w, tid)
+  
+        case traverse id rs of
+          Left a -> do
+            io $ forM_ rs $ \r -> case r of
+              Right (Right (_, tid)) -> killThread tid
```

Comment:
Might be nicer to use an existing library like [async](https://hackage.haskell.org/package/async) to take care of cleaning up threads.

Reply:
`async` would indeed be nice, though it also depends on `stm`, which would effectively negate the benefits of ditching it in the first place.

However, I see now that `race`/`concurrently` might not use `stm` at all, and if so that'd be a much cleaner way of doing things. Will dig into it a bit more.